src/pages/api/accounts-data.ts

```diff
@@ -0,0 +1,184 @@
+import { getSubsocialApi } from '@/subsocial-query/subsocial/connection'
+import { MinimalUsageQueueWithTimeLimit } from '@/utils/data-structure'
+import axios from 'axios'
+import { request } from 'graphql-request'
+import gql from 'graphql-tag'
+
+import { NextApiRequest, NextApiResponse } from 'next'
+import { z } from 'zod'
+
+export type AccountData = {
+  grillAddress: string
+  evmAddress: string
+  ensName: string | null
+  withEnsAvatar: boolean
+}
+
+const querySchema = z.object({
+  addresses: z.array(z.string()),
+})
+
+export type ApiPostsParams = z.infer<typeof querySchema>
+
+export type ApiAccountDataResponse = {
+  success: boolean
+  message: string
+  errors?: any
+  data?: any
+  hash?: string
+}
+
+const MAX_CACHE_ITEMS = 500_000
+const accountsDataCache = new MinimalUsageQueueWithTimeLimit<AccountData>(
+  MAX_CACHE_ITEMS,
+  5
+)
+
+const GET_ENS_NAMES = gql(`
+  query GetEnsNames($evmAddresses: [String!]) {
+    domains(where: { resolvedAddress_: { id_in: $evmAddresses } }) {
+      name
+      resolvedAddress {
+        id
+      }
+    }
+  }
+`)
+
+export default async function handler(
+  req: NextApiRequest,
+  res: NextApiResponse<ApiAccountDataResponse>
+) {
+  const query = req.query.addresses
+  const params = querySchema.safeParse({
+    addresses: Array.isArray(query) ? query : [query],
+  })
+
+  if (!params.success) {
+    return res.status(400).send({
+      success: false,
+      message: 'Invalid request body',
+      errors: params.error.errors,
+    })
+  }
+
+  const accountsData = await getAccountsDataFromCache(
+    params.data.addresses,
+    req.method || 'GET'
```

Comment:
I think its better to have the if for method in here, rather in `getAccountsDataFromCache`
so you also don't need to pass the method to that function